### PR TITLE
Switch reduce_scatter and all_gather in DeviceMesh to use functional collectives

### PR DESCRIPTION
**Pull request by kumpera**
Among the changes is the introduction of gather_dim and scatter_dim in DeviceMesh collectives to simplify user code.

The current plan is to keep padding and gather/scatter dim support in DeviceMesh while we explore  optimization opportunities in Inductor.

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: aten/src/ATen/native/Collectives.cpp [hunk 1]
```diff
@@ -24,7 +24,7 @@ at::Tensor all_gather_into_tensor(at::Tensor const& shard, const c10::string_vie
     TORCH_INTERNAL_ASSERT(false);
 }
 
-at::Tensor reduce_scatter_tensor(at::Tensor const& input, const c10::string_view reduceOp, int64_t scatter_dim, const c10::string_view tag, c10::ArrayRef<int64_t> ranks, int64_t group_size) {
+at::Tensor reduce_scatter_tensor(at::Tensor const& input, const c10::string_view reduceOp, const c10::string_view tag, c10::ArrayRef<int64_t> ranks, int64_t group_size) {
```
Comment (wanchaol) [inline aten/src/ATen/native/Collectives.cpp:27]: curious why we delete the scatter_dim? I think it's still valuable to do the transpose thing in the  func collective level? We can do it in func collective python level

File: torch/distributed/_tensor/device_mesh.py [hunk 2]
```diff
@@ -408,12 +414,49 @@ def all_gather(
...
                 mesh dimension as source of truth.
+            gather_dim (int, optional): Dimension to concatenate the resulting tensor.
+            gather_size (torch.Size, optional): Desired shape of the gathered tensor.
 
+            This suppport up to 1 elem of padding on gather_dim.
         Returns:
-            A :class:`Work` object
+            A :class:`torch.Tensor` object
         """
+        my_coordinate = self.get_coordinate()
```
Comment (wanchaol) [inline torch/distributed/_tensor/device_mesh.py:424]: could we move the padding logic back to placement (sharding) types? I feel this is a dtensor sharding algorithm but not related to collective itself. A clean contract between sharding and collective would actually help other cases, and we might want to change our sharding/padding logic later as we discussed.

Comment (kumpera) [reply torch/distributed/_tensor/device_mesh.py:424]: Let me try that, but I believe they don't compose so the result will be worse.

File: aten/src/ATen/native/Collectives.cpp (hunk 1 above)
Comment (kumpera) [reply aten/src/ATen/native/Collectives.cpp:27]: reverting this to reduce the CI churn

File: torch/distributed/_tensor/device_mesh.py (hunk 2 above)
Comment (kumpera) [reply torch/distributed/_tensor/device_mesh.py]: We decided to have dim gather/split separate from padding logic.

File: torch/distributed/_tensor/device_mesh.py [hunk 3]
```diff
@@ -398,75 +396,51 @@ def all_reduce(
...
+            flat_tensor = funcol.all_reduce(input, reduceOp=op_name, group=dim_group)
+            chunks = flat_tensor.chunk(get_world_size(dim_group))
+            scatter_tensor = chunks[get_rank(dim_group)]
         else:
             raise RuntimeError(
                 f"backend {self._backend} does not support reduce_scatter!"
             )
-        return fut
+
+        # if needs_padding:
```
Comment (wanchaol) [inline torch/distributed/_tensor/device_mesh.py:439]: nit: delete this as we did this in the placements already?

**Review by wanchaol — APPROVED** [review]
Great work! Thanks for doing this! I only have some nits

Looks like there're additional tests need to be fixed according to CI https://hud.pytorch.org/pr/96226#12342038749

File: torch/distributed/_functional_collectives.py [hunk 4]
```diff
@@ -277,6 +282,38 @@ def all_reduce(self: torch.Tensor, reduceOp: str, group: RANK_TYPES, tag: str =
...
+        ProcessGroup: Will perform a collective using the ranks and tag of the PG.
+        DeviceMesh: Do a SPMD collective over all ranks of the mesh
+        (DeviceMesh, int): Do a MPMD collective over one dimension of the DeviceMesh
+
+    :: N.B. If you pass a PG or a 1D list to perform a MPMD collective, the compiler won't be able to recover
+    that information and perform collective algebraic optimization. Use other forms of input for that.
+    """
+    assert self.is_contiguous()
+    tag, rankset, group_size = _expand_group(group, tag)
+    tensor = torch._C._nn.all_gather_into_tensor(self, tag, rankset, group_size)  # type: ignore[attr-defined]
```
[the remark this answers is not in the document]
Comment (kumpera) [reply torch/distributed/_functional_collectives.py:309]: Makes sense, the naming needs fixing, I don't want to change on that part of the code until https://github.com/pytorch/pytorch/pull/97793 lands though.

**Comment by yanbing-j** [comment]
Hi @kumpera , I think the change of submodule `third_party/ideep` in this PR is not intentional, could you please revert this PR to remove the change of submodule? Thanks!

**Comment by kumpera** [comment]
> Hi @kumpera , I think the change of submodule `third_party/ideep` in this PR is not intentional, could you please revert this PR to remove the change of submodule? Thanks!

that was indeed a mistake. ouch :/

**Pull request closed without merging**